tests/backends/vllm/test_migration.py

```diff
@@ -0,0 +1,137 @@
+# Copyright (c) 2024, Alibaba Group;
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+import torch
+import time
+import ray
+from ray.util.queue import Queue as RayQueue
+from ray.util.scheduling_strategies import NodeAffinitySchedulingStrategy
+
+from vllm import EngineArgs, SamplingParams
+from vllm.utils import random_uuid
+
+from llumnix.backends.vllm.llm_engine import BackendVLLM
+from llumnix.llumlet.llumlet import Llumlet
+from llumnix.backends.utils import BackendType
+from llumnix.config import MigrationConfig
+from llumnix.server_info import ServerInfo
+
+from .test_llm_engine import MockEngine
+from .utils import create_dummy_prompt
+
+TEST_PROMPTS = ["hello world, ",
+                "Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n",
+                "Write a short story about a robot that dreams for the first time.\n",
+                "Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n",
+                "Swahili: 'The early bird catches the worm.'\n"]
+
+class MockBackendVLLM(BackendVLLM):
+    def __init__(self):
+        self.engine = MockEngine()
+
+class MockLlumlet(Llumlet):
+    def __init__(self):
+        self.instance_id = "0"
+        self.backend_engine = MockBackendVLLM()
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2,
+                    reason="Need at least 2 GPUs to run the test.")
+def test_migration_correctness():
+    ray.init(namespace="llumnix", ignore_reinit_error=True)
+    engine_args = EngineArgs(model="facebook/opt-125m",worker_use_ray=True)
+    id_rank_map = {"0":0,"1":1}
+    migration_config = MigrationConfig("LCFS", "gloo",16,1,4,4,5,False)
+    que = RayQueue(actor_options={
+        "scheduling_strategy": NodeAffinitySchedulingStrategy(
+            node_id=ray.get_runtime_context().get_node_id(),
+            soft=False,)
+    })
+    server_info = ServerInfo("0",que)
+
+    llumlet_0:Llumlet = Llumlet.from_args(True,
+                      "0",
```

Comment:
Why do we need to create a llumlet in the VLLM engine test? Is this a llumlet test?

Reply:
It's an e2e test to ensure migration correctness.